### PR TITLE
Create super-resolution demo for TensorFlow

### PR DESCRIPTION
**Pull request by AzizZayed**
## Description ##

This is a demo of an image enhancement GAN model called ESRGAN for TensorFlow.

### CODE REVIEW

File: examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        return output.split(sampleSize)
+                .stream()
+                .map(array -> ImageFactory.getInstance().fromNDArray(array.squeeze()))
+                .toArray(Image[]::new);
+    }
+
+    @Override
+    public Batchifier getBatchifier() {
+        return null;
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java:51]:
```suggestion
        return Batchifier.STACK;
```

File: examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import ai.djl.ndarray.NDArrays;
+import ai.djl.ndarray.NDList;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.translate.Batchifier;
+import ai.djl.translate.Translator;
+import ai.djl.translate.TranslatorContext;
+import java.util.Arrays;
+
+public class SRTranslator implements Translator<Image[], Image[]> {
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java:27]: We can leverage Batchifier
```suggestion
public class SuperResultionTranslator implements Translator<Image, Image> {
```

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                        .optModelUrls(modelUrl)
+                        .optOption("Tags", "serve")
+                        .optOption("SignatureDefKey", "serving_default")
+                        .optTranslator(new SRTranslator())
+                        .optProgress(new ProgressBar())
+                        .build();
+
+        try (ZooModel<Image[], Image[]> model = criteria.loadModel();
+                Predictor<Image[], Image[]> enhancer = model.newPredictor()) {
+            return enhancer.predict(inputImages);
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:117]:
```suggestion
            return enhancer.batchPredict(inputImages);
```

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            stitches.add(NDArrays.concat(new NDList(left, right), 1));
+        }
+
+        return stitches.stream()
+                .map(array -> array.toType(DataType.UINT8, false))
+                .map(array -> ImageFactory.getInstance().fromNDArray(array))
+                .toArray(Image[]::new);
+    }
+
+    public static Image[] enhance(Image[] inputImages)
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:97]:
```suggestion
    public static List<Image> enhance(List<Image> inputImages)
```

File: examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java (hunk 2 above)
Comment (frankfliu) [reply examples/src/main/java/ai/djl/examples/inference/sr/SRTranslator.java:27]: use full name or `SrTranslator`, see: https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    private static void save(Image[] images, String name, Path path) throws IOException {
+        for (int i = 0; i < images.length; i++) {
+            Path imagePath = path.resolve(name + i + ".png");
+            images[i].save(Files.newOutputStream(imagePath), "png");
+        }
+    }
+
+    private static Image[] group(Image[] input, Image[] generated) {
+        NDManager manager = Engine.getInstance().newBaseManager();
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:76]: try (NDManager manager = ...) {
}

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    private static List<Image> group(List<Image> input, List<Image> generated) {
+        NDList stitches = new NDList(input.size());
+
+        try (NDManager manager = Engine.getInstance().newBaseManager()) {
+            for (int i = 0; i < input.size(); i++) {
+                int width = input.get(i).getWidth();
+                int height = input.get(i).getHeight();
+
+                NDArray left = input.get(i).toNDArray(manager);
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:90]: I think we should move stitch part into translator

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+    private SuperResolution() {}
+
+    public static void main(String[] args) throws ModelException, TranslateException, IOException {
+        String imagePath = "src/test/resources/";
+        ImageFactory imageFactory = ImageFactory.getInstance();
+
+        List<Image> inputImages =
+                Arrays.asList(
+                        imageFactory.fromFile(Paths.get(imagePath + "fox.png")),
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:53]: Can you just link to original fox and monkey image?

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java (hunk 6 above)
Comment (AzizZayed) [reply examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:90]: Stitching the input image and output image is just for demo purposes so we can see the difference between the input and output. Do you still think this should go in the translator? It isn't technically part of postprocessing.

File: examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java (hunk 7 above)
Comment (AzizZayed) [reply examples/src/main/java/ai/djl/examples/inference/sr/SuperResolution.java:53]: I just added them to the repo temporarily. Just for the review. I modified the images however, so I think we should put them in S3.

**Review by frankfliu — APPROVED** [review]

**Pull request merged**